# Thông tin
- Họ tên: Nguyễn Phúc Thuần
- MSSV: 20120380

# Bài làm

## 4.2

$P=(-1,4), Q=(2,5)$

$E: Y^2 = X^3 + 17$

### a

$PQ = \dfrac{1}{3} \cdot x + \dfrac{13}{3}$

$R = PQ \cap E$

$\Rightarrow (\dfrac{1}{3} \cdot x + \dfrac{13}{3})^2 = X^3 + 17$

$\Rightarrow X^3 + 17 - (\dfrac{1}{3} \cdot x + \dfrac{13}{3})^2 = 0$

$\Rightarrow X^3 + 17 - \dfrac{1}{9} \cdot x^2 - 2\dfrac{13}{9} \cdot x - \dfrac{13}{3}^2 = 0$

Ta đã biết $X = -1$ và $X = 2$ là 2 nghiệm của phương trình trên

Từ đó dễ dàng suy ra nghiệm còn lại là $X = \dfrac{-8}{9}$

Suy ra $R = (\dfrac{-8}{9},\dfrac{109}{27})$

Vậy $P \oplus Q = R'(\dfrac{-8}{9},\dfrac{-109}{27})$





$P \ominus Q = P \oplus (-Q) = P \oplus Q'$ với $Q'(2, -5)$

$PQ' = -3 \cdot X + 1$

$R = PQ' \cap E$

$\Rightarrow (-3 \cdot X + 1)^2 = X^3 + 17$

$\Rightarrow (-3 \cdot X + 1)^2 - X^3 - 17 = 0$

Ta tính được nghiệm thứ ba $X =8$

Suy ra $R(8, -23)$

Vậy $P \ominus Q = R'(8,23)$

### b

Ta có: $\dfrac{dY}{dX} = \dfrac{3X^2}{2Y}$

$\Rightarrow \lambda_P = \dfrac{3}{8} 
\Rightarrow \text{ tiếp tuyến tại P là } 
L:Y=\dfrac{3}{8} \cdot X + \dfrac{35}{8} $

Khi đó:

$(\dfrac{3}{8} \cdot X + \dfrac{35}{8})^2 = X^3 + 17$
$\Rightarrow X = \dfrac{137}{64}$ là nghiệm thứ 3 cần tìm

Do đó $2P = (\dfrac{137}{64}, -\dfrac{2651}{512})$

Tương tự:
$\dfrac{dY}{dX} = \dfrac{3X^2}{2Y}$
$\Rightarrow \lambda_Q = \dfrac{6}{5} 
\Rightarrow \text{ tiếp tuyến tại Q là } 
L:Y=\dfrac{6}{5} \cdot X + \dfrac{13}{5} $

Khi đó:

$(\dfrac{6}{5} \cdot X + \dfrac{13}{5})^2 = X^3 + 17$
$\Rightarrow X = \dfrac{-64}{25}$ là nghiệm thứ 3 cần tìm

Do đó $2Q = (\dfrac{-64}{25}, -\dfrac{59}{125})$




## 5.3

Khai triển vế phải, ta được đẳng thức:

$X^3 + AX + B = X^3 - (e_1 + e_2 + e_3)X^2 + (e_1e_2 + e_1e_3 + e_2e_3)X - e_1e_2e_3$

Đồng nhất các hệ số:

$
e_1 + e_2 + e_3 = 0 \\
e_1e_2 + e_1e_3 + e_2e_3 = A \\
e_1e_2e_3 = B \\
$

Giả sử $e_2 = e_3$. Khi đó:

$
\begin{align}
e_1 + 2e_2 = 0 \tag{1} \\
2e_1e_2 + {e_2}^2 = A \tag{2} \\
e_1{e_2}^2 = B \tag{3}
\end{align}
$

Từ $(1)$ thế $e_1 = -2e_2$ vào $(2)$ và $(3)$ ta được:
$-3{e_2}^2 = A$ và $-2{e_2}^3 = B$

Khi đó $4A^3 + 27B^2 = 4(-3{e_2}^2)^3 + 27(-2{e_2}^3)^2 = 0$

## 5.6

In [624]:
def inverse(x, p):
    if x % p == 0:
        return None
    
    return pow(x, p-2, p)
    
def ellyptic_addition(P, Q, E, p):
    __O__ = (0,0)
    if P == __O__:
        return Q
    
    if Q == __O__:
        return P
    
    x1, y1 = P
    x2, y2 = Q
    
    if x1 == x2 and y1 == -y2:
        return __O__
    
    if P != Q:
        denom = inverse(x2 - x1, p)
        if denom == None:
            return __O__
        
        __lambda__ = ((y2 - y1)*denom) % p
    else:
        A,_ = E
        denom = inverse(2*y1, p)
        if denom == None:
            return __O__
        
        __lambda__ = ((3*pow(x1, 2) + A) * denom) % p
        
    x3 = pow(__lambda__, 2) - x1 - x2
    y3 = __lambda__*(x1 - x3) - y1
    
    return (x3 % p, y3 % p)

def square_root(x, p):
    return [i for i in range(0,p) if pow(i, 2, p) == x]

def elliptic_points(a, b, p):
    E_points = [(0,0)]
    
    for x in range(0,p):
        yy = pow(x, 3) + a*x + b
        sqr = square_root(yy % p, p)
    
    if len(sqr) > 0:
        for y in sqr:
            E_points.append((x, y))
            
    return E_points
    
def addition_table_for_ellyptic_curve(a, b, p):
    
    E = (a,b)
    
    E_points = elliptic_points(a, b, p)
    
    print(f'E(Fp) = {E_points}')
    
    table = []
    
    for P in E_points:
        row = []
        for Q in E_points:
            row.append(ellyptic_addition(P, Q, E, p))
            
        table.append(row)
    
    return table

def print_elliptic_addition_table(table):   
    first_row = table[0]
    row_string = ""
    for point in first_row:
            if point == (0,0):
                row_string += '\tO\t'
            else:
                row_string += f'\t{point}\t'
    print(f'\t|{row_string}')
    
    for row in table:
        row_string = ""
        for point in row:
            if point == (0,0):
                row_string += '\tO\t'
            else:
                row_string += f'\t{point}\t'
        print(f'{row[0]}\t| {row_string}')
    

### (a)

In [625]:
table = addition_table_for_ellyptic_curve(1, 2, 5)
print_elliptic_addition_table(table)

E(Fp) = [(0, 0), (4, 0)]
	|	O		(4, 0)	
(0, 0)	| 	O		(4, 0)	
(4, 0)	| 	(4, 0)		O	


### (b)

In [626]:
table = addition_table_for_ellyptic_curve(2, 3, 7)
print_elliptic_addition_table(table)

E(Fp) = [(0, 0), (6, 0)]
	|	O		(6, 0)	
(0, 0)	| 	O		(6, 0)	
(6, 0)	| 	(6, 0)		O	


### (c)

In [627]:
table = addition_table_for_ellyptic_curve(2, 5, 11)
print_elliptic_addition_table(table)

E(Fp) = [(0, 0)]
	|	O	
(0, 0)	| 	O	


## 5.9

Giả sử $n = is + r$ với $0 \le r < s$, khi đó:

$Q = nP = (is + r)P = i\cdot(sP) + rP = iO + rP = rP$

Vì $n_0$ là nghiệm nhỏ nhất của phương trình $Q = nP$ nên $r \ge n_0$

Với $r = n_0 \Rightarrow n = 0s + r = 0s + n_0 = n_0$ (hiển nhiên đúng)

Với $r > n \Rightarrow rP - n_0P = (r-n_0)P = Q - Q = O \Rightarrow r - n_0 \ge s$

Điều này trái với giả thiết ban đầu là $r < s$

Do đó khả năng duy nhất xảy ra là $r = n_0$ hay $n = is + n_0$ (điều phải chứng minh)


## 5.10

In [628]:
def double_and_add(P, n, a, b, p):
    
    Q = P
    R = (0,0)
    
    while n > 0:
        if n & 1:
            R = ellyptic_addition(R, Q, (a, b), p)
        Q = ellyptic_addition(Q, Q, (a, b), p)
        n = n >> 1
    
    return R

### (a)

In [629]:
a = 23
b = 13
p = 83
P = (24, 14)
n = 19

nP = double_and_add(P, n, a, b, p)
print(f'{n}P = {nP}')


19P = (24, 69)


### (b)

In [630]:
a = 143
b = 367
p = 613
P = (195, 9)
n = 23

nP = double_and_add(P, n, a, b, p)
print(f'{n}P = {nP}')

23P = (485, 573)


### (c)

In [631]:
a = 1828
b = 1675
p = 1999
P = (1756, 348)
n = 11

nP = double_and_add(P, n, a, b, p)
print(f'{n}P = {nP}')

11P = (1068, 1540)


### (d)

In [632]:
a = 1541
b = 1335
p = 3221
P = (2898, 439)
n = 3211

nP = double_and_add(P, n, a, b, p)
print(f'{n}P = {nP}')

3211P = (243, 1875)


## 5.11

In [633]:
def binary_presentation(n: int):
    result = []
    
    while n != 0:
        result.append(n & 1)
        n = n >> 1
    
    # result.reverse()    
    return result

In [634]:
def ternary_presentation(n: int):
    binary_present = binary_presentation(n)
    binary_present.append(0)
    
    terna_present = []
    sub_terna = []
    
    for s, bit in enumerate(binary_present):
        if bit == 1:
            sub_terna.append(s)
        elif bit == 0 and len(sub_terna) > 0:
            start = sub_terna[0]
            t = len(sub_terna)
            
            if (-1, start) in terna_present:
                terna_present.remove((-1, start))
                terna_present.append((-1, start+1))
            else:
                terna_present.append((-1, start))
            
            if (1, start+t) in terna_present:
                terna_present.remove((1, start+t))
                terna_present.append((1, start+t+1))
            else:
                terna_present.append((1, start+t))
                
            sub_terna.clear()

    for (sign, expo) in terna_present:
        if (-sign, expo) in terna_present:
            terna_present.remove((sign, expo))
            terna_present.remove((-sign, expo))

    terna_present.sort(key = lambda x:x[1])
    return terna_present

def print_binary_present(n, binary_present):
    present_string = ""
    for expo, bit in enumerate(binary_present):
        if bit == 1:
            present_string += f'2^{expo} + '
    
    print(f'Binary present: {n} = {present_string[:-2]}')
    
def print_ternary_present(n, ternary_present):
    present_string = ""
    for sign, expo in ternary_present:
        if sign == 1:
            present_string += '+ '
        if sign == -1:
            present_string += '- '
        
        present_string += f'2^{expo} '
    
    print(f'Tenary present: {n} = {present_string}')

In [635]:
n = 349
print_binary_present(n, binary_presentation(n))
print_ternary_present(n, ternary_presentation(n))


Binary present: 349 = 2^0 + 2^2 + 2^3 + 2^4 + 2^6 + 2^8 
Tenary present: 349 = - 2^0 + 2^1 - 2^2 + 2^5 - 2^6 + 2^7 - 2^8 + 2^9 


In [636]:
n = 9337
print_binary_present(n, binary_presentation(n))
print_ternary_present(n, ternary_presentation(n))

Binary present: 9337 = 2^0 + 2^3 + 2^4 + 2^5 + 2^6 + 2^10 + 2^13 
Tenary present: 9337 = - 2^0 + 2^1 - 2^3 + 2^7 - 2^10 + 2^11 - 2^13 + 2^14 


In [637]:
n = 38728
print_binary_present(n, binary_presentation(n))
print_ternary_present(n, ternary_presentation(n))

Binary present: 38728 = 2^3 + 2^6 + 2^8 + 2^9 + 2^10 + 2^12 + 2^15 
Tenary present: 38728 = - 2^3 + 2^4 - 2^6 + 2^7 - 2^8 + 2^11 - 2^12 + 2^13 - 2^15 + 2^16 


In [638]:
n = 8379483273489
print_binary_present(n, binary_presentation(n))
print_ternary_present(n, ternary_presentation(n))

Binary present: 8379483273489 = 2^0 + 2^4 + 2^8 + 2^11 + 2^12 + 2^13 + 2^15 + 2^16 + 2^17 + 2^18 + 2^19 + 2^20 + 2^32 + 2^33 + 2^34 + 2^35 + 2^36 + 2^39 + 2^40 + 2^41 + 2^42 
Tenary present: 8379483273489 = - 2^0 + 2^1 - 2^4 + 2^5 - 2^8 + 2^9 - 2^11 + 2^14 - 2^15 + 2^21 - 2^32 + 2^37 - 2^39 + 2^43 


## 5.13

### a
Bob gửi cho Alice $Q_B = n_BP = (1432, 667)$

In [639]:
nB = 1943
P = (1980, 431)
a = 171
b = 853
p = 2671
print('Q =',double_and_add(P, nB, a, b, p))

Q = (1432, 667)


### c
Bằng máy tính, khi Eve biết được $Q_A$ mà Alice gửi cho Bob và $P$, Alice có thể dùng nó để tính ra $n_A$ bằng Bruteforce nếu $p$ không quá lớn

In [640]:
def find_nA(Q, P, a, b, p):
    for nE in range(2, p):
        if Q == double_and_add(P, nE, a, b, p):
            return nE

QA = (2110, 543)
print(f'secrect number of Alice is: {find_nA(QA, P, a, b, p)}')

secrect number of Alice is: 726


Như ta thấy, với đoạn chương trình đơn giản trên, Eve có thể tìm ra $n_A = 726$ chỉ trong một vài giây.

### d
Bob tính được $Q_B = (161, 240)$, và gửi cho Alice $x_B = 161$

Để tính giá trị chung, Bob cần tính lại $Q_A$ và có thể tính một cách dễ dàng bằng cách thế vào phương trình của đường cong Elliptic:

$y^2_A = x^3_A + 171x_A + 853 = 1203 \Rightarrow y_A = 2575$ (mod $2671$)

Khi đó key chung sẽ là: $n_BQ_A = 875\cdot (2, 2575) = (1708, 1419)$ và giá trị được dùng chung là $x = 1708$